File: Ravi/Microsoft/Day 3/MostProfitablePathinaTree.java
class Solution {
    
    int ans = Integer.MIN_VALUE;
    
    boolean dfs(int node, ArrayList<Integer> adj[], int[] vis, int[] dis, int curr) {
        
        vis[node] = 1;
        
        if(node == 0) {
            dis[node] = curr;    
            return true;
        }
        
        boolean flag = false;
        
        for(int i : adj[node]) {
            if(vis[i] == 0) {
                if(dfs(i, adj, vis, dis, curr + 1)) {
                    flag = true;
                    break;
                }
            }
        }
        
        if(flag) {
            dis[node] = curr;
            return true;
        }
        
        return false;
        
    }
    
    void dfs2(int node, ArrayList<Integer> adj[], int[] vis, int[] dis, int curr, int cost, int[] amount) {
        
        vis[node] = 1;
        
        if(dis[node] > curr) cost += amount[node];
        else if(dis[node] == curr) cost += (amount[node] / 2);
        
        for(int i : adj[node]) {
            if(vis[i] == 0) dfs2(i, adj, vis, dis, curr + 1, cost, amount);
        }
        
        if(node != 0 && adj[node].size() == 1) ans = Math.max(ans, cost);
        
    }
    
    public int mostProfitablePath(int[][] edges, int bob, int[] amount) {
        
        int n = amount.length;
        ArrayList<Integer> adj[] = new ArrayList[n];
        for(int i = 0; i < n; i++) adj[i] = new ArrayList<>();  
        
        for(int[] e : edges) {
            adj[e[0]].add(e[1]);
            adj[e[1]].add(e[0]);
        }
        
        int[] dis = new int[n];
        for(int i = 0; i < n; i++) dis[i] = (int)(1e8);
        
        int[] vis = new int[n];
        
        dfs(bob, adj, vis, dis, 0);

        vis = new int[n];
        dfs2(0, adj, vis, dis, 0, 0, amount);
        
        return ans;
        
    }
}

File: Nehal/Microsoft/Day 3/most-profitable-path-in-a-tree.cpp
class Solution
{
public:
    bool DFS(int src, int time, unordered_map<int, int> &path, vector<bool> &visited, vector<vector<int>> &graph)
    {
        path[src] = time;
        visited[src] = true;
        if (src == 0)
        {
            return true;
        }
        for (auto adj : graph[src])
        {
            if (!visited[adj])
            {
                if (DFS(adj, time + 1, path, visited, graph))
                {
                    return true;
                }
            }
        }
        path.erase(src);
        return false;
    }

    int mostProfitablePath(vector<vector<int>> &edges, int bob, vector<int> &amount)
    {
        int n = edges.size() + 1;
        vector<vector<int>> graph(n);
        for (auto it : edges)
        {
            graph[it[0]].push_back(it[1]);
            graph[it[1]].push_back(it[0]);
        }

        // Find the path taken by bob to reach zero along with the time at which bob visited that node
        unordered_map<int, int> path;
        vector<bool> visited(n, false);
        DFS(bob, 0, path, visited, graph);

        // Push {src, time, income}
        queue<vector<int>> q;
        q.push({0, 0, 0});
        visited.assign(n, false);

        int ans = INT_MIN;
        while (!q.empty())
        {
            int src = q.front()[0], time = q.front()[1], income = q.front()[2];
            q.pop();
            visited[src] = true;

            // If bob didn't visit this node
            if (path.find(src) == path.end())
            {
                income += amount[src];
            }
            else
            {
                // Alice visits it first
                if (time < path[src])
                {
                    income += amount[src];
                }
                // Both visit at the same time
                else if (time == path[src])
                {
                    income += (amount[src] / 2);
                }
            }

            // Updating when it is leaf
            if (graph[src].size() == 1 && src != 0)
            {
                ans = max(ans, income);
            }
            // Exploring adjacent vertices
            for (auto adj : graph[src])
            {
                if (!visited[adj])
                {
                    q.push({adj, time + 1, income});
                }
            }
        }

        return ans;
    }
};